game.py

```diff
@@ -0,0 +1,908 @@
+from math import inf
+from copy import copy, deepcopy
+import random
+import pickle
+import time
+
+class GameState():
+    
+    def __init__(self, board, player_next, connect_number, number_of_moves, moves_list = []):
+        self._board = board
+        self._player_next = player_next
+        self._connect_number = connect_number
+        self._number_of_moves = number_of_moves
+        self._moves_list = moves_list
+        # self._last_move = last_move
+        # self._previous_game_state = previous_game_state
+
+    def _copy(self):
+        board = deepcopy(self._board)
+        player_next = deepcopy(self._player_next)
+        connect_number = deepcopy(self._connect_number)
+        number_of_moves = deepcopy(self._number_of_moves)
+        moves_list = deepcopy(self._moves_list)
+        # previous_game_state = deepcopy(self)
+        return GameState(board, player_next, connect_number, number_of_moves, moves_list)
+                
+    def get_board(self):

```

Comment:
This is quite in-depth, including comparisons to other modern-but-less-flexible languages, but is a good resource when thinking about getters and setters: https://realpython.com/python-getter-setter/

Another more-recent option is to have this class be a `@dataclass`. This concept is still new and evolving in Python, but is another attempt to remove some of the grunt-work here. But really, the options here quickly multiply and currently are a bit confusing. This is a long article talking about how to work with both properties and dataclasses: https://florimond.dev/en/posts/2018/10/reconciling-dataclasses-and-properties-in-python/

Bringing it back to your class, as is you want:
* The class state to be mutable. i.e. methods like `make_move` change internal state and return nothing (void), rather than returning new instances of the class (and leaving the old one untouched).
* No class state to be mutated by code outside of the class
* Code outside that class has read-access to much of the state e.g. the board, next player to move etc.

[Zen of Python](https://peps.python.org/pep-0020/) hopefully states that for any situation `There should be one-- and preferably only one --obvious way to do it.` In your case, I think this would be:
* Keep the constructor the same (I think making it a dataclass in this case doesn't end up saving any code if we want to keep the fields private, and just increases complexity)
* Make all of these getters `@property` methods instead